routing/coding.hpp

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include "routing/routing_exceptions.hpp"
+
+#include "coding/bit_streams.hpp"
+#include "coding/elias_coder.hpp"
+
+#include "base/assert.hpp"
+#include "base/bits.hpp"
+#include "base/checked_cast.hpp"
+#include "base/macros.hpp"
+
+#include <limits>
+#include <type_traits>
+
+namespace routing
+{
+template <typename T, typename Source>
+T ReadDelta(BitReader<Source> & reader)
+{
+  static_assert(std::is_unsigned<T>::value, "T should be an unsigned type");
+  uint64_t const decoded = coding::DeltaCoder::Decode(reader);
+  if (decoded > std::numeric_limits<T>::max())
+  {
+    MYTHROW(CorruptedDataException,
+            ("Decoded value", decoded, "out of limit", std::numeric_limits<T>::max()));
+  }
+
+  return static_cast<T>(decoded);
+}
+
+template <typename T, typename Sink>
+void WriteDelta(BitWriter<Sink> & writer, T value)
+{
+  static_assert(std::is_unsigned<T>::value, "T should be an unsigned type");
+  ASSERT_GREATER(value, 0, ());
+
+  bool const success = coding::DeltaCoder::Encode(writer, static_cast<uint64_t>(value));
+  ASSERT(success, ());
+  UNUSED_VALUE(success);
+}
+
+template <typename T, typename Source>
+T ReadGamma(BitReader<Source> & reader)
+{
+  static_assert(std::is_unsigned<T>::value, "T should be an unsigned type");
+  uint64_t const decoded = coding::GammaCoder::Decode(reader);
+  if (decoded > std::numeric_limits<T>::max())
+  {
+    MYTHROW(CorruptedDataException,
+            ("Decoded value", decoded, "out of limit", std::numeric_limits<T>::max()));
+  }
+
+  return static_cast<T>(decoded);
+}
+
+template <typename T, typename Sink>
+void WriteGamma(BitWriter<Sink> & writer, T value)
+{
+  static_assert(std::is_unsigned<T>::value, "T should be an unsigned type");
+  ASSERT_GREATER(value, 0, ());
+
+  bool const success = coding::GammaCoder::Encode(writer, static_cast<uint64_t>(value));
+  ASSERT(success, ());
+  UNUSED_VALUE(success);
+}
+
+template <typename T, typename SignedT = typename std::make_signed<T>::type>
+SignedT ModuleCast(T value)
+{
+  static_assert(std::is_unsigned<T>::value, "T should be an unsigned type");
+
+  if (value <= static_cast<T>(std::numeric_limits<SignedT>::max()))
+    return static_cast<SignedT>(value);
+
+  auto constexpr minSignedT = std::numeric_limits<SignedT>::min();
+  return static_cast<SignedT>(value - static_cast<T>(minSignedT)) + minSignedT;
+}
+
+// Encodes current as delta compared with prev.
+template <typename T, typename UnsignedT = typename std::make_unsigned<T>::type>
```

Comment:
Why not to move UnsignedT to the body of the function, near type alias for SignedT.